### PR TITLE
Make context propagation requirements explict

### PR DESCRIPTION
**Pull request by joaopgrassi**
This PR is an initial attempt in bring the contents from [OTEP 0205](https://github.com/open-telemetry/oteps/blob/main/text/trace/0205-messaging-semantic-conventions-context-propagation.md) into the existing messaging semantic conventions. 

As stated in the OTEP, the existing messaging semantic conventions rely on context to be propagated via the message but this requirement is implicit and not clearly stated in the conventions today.

The OTEP's and this PR intention is to make this requirement explicit in the current conventions, without introducing breaking changes or new behavior.

### CODE REVIEW

File: specification/trace/semantic_conventions/messaging.md [hunk 1]
```diff
@@ -80,6 +80,38 @@ Often such destinations are unnamed or have an auto-generated name.
 
 Given these definitions, the remainder of this section describes the semantic conventions for Spans describing interactions with messaging systems.
 
+### Context propagation
```
Comment (joaopgrassi) [inline specification/trace/semantic_conventions/messaging.md:83]: I added this new section at the top, since I believe this is an important topic for whoever is consuming this document.

File: specification/trace/semantic_conventions/messaging.md [hunk 2]
```diff
@@ -80,6 +80,48 @@ Often such destinations are unnamed or have an auto-generated name.
...
+
+A message may pass many different components and layers in one or more intermediaries
+when it is propagated from the producer to the consumer(s). To be able to correlate
+consumer traces with producer traces using the existing context propagation mechanisms,
+all components must propagate context down the chain.
+
+This imposes a difficulty because it cannot be assumed, and in many cases,
+it is not even desired, that all components and layers to be instrumented and
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:91]:
```suggestion
it is not even desired, that all components and layers are instrumented and
```

File: specification/trace/semantic_conventions/messaging.md [hunk 3]
```diff
@@ -80,6 +80,48 @@ Often such destinations are unnamed or have an auto-generated name.
...
+propagate context according to OpenTelemetry requirements.
+
+To be able to correlate consumer traces with producer traces without requiring
+intermediary instrumentation, the context needs to be propagated on a
+*per-message* basis instead of on a *per-request* basis. This allows all components
+to have access to the same per-message context information, making it possible
+to correlate all the stages involved in processing a message with the message's
+creation.
+
+A message *creation context* allows correlating producer with consumer(s)
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:101]:
```suggestion
A message *creation context* allows correlating producers with consumers
```
or 
```suggestion
A message *creation context* allows correlating a producer with one or more consumers
```

File: specification/trace/semantic_conventions/messaging.md [hunk 4]
```diff
@@ -80,6 +80,48 @@ Often such destinations are unnamed or have an auto-generated name.
 
 Given these definitions, the remainder of this section describes the semantic conventions for Spans describing interactions with messaging systems.
 
+### Context propagation
+
+A message may pass many different components and layers in one or more intermediaries
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:85]:
```suggestion
A message may traverse many different components and layers in one or more intermediaries
```

File: specification/trace/semantic_conventions/messaging.md [hunk 5]
```diff
@@ -80,6 +80,48 @@ Often such destinations are unnamed or have an auto-generated name.
...
+when it is propagated from the producer to the consumer(s). To be able to correlate
+consumer traces with producer traces using the existing context propagation mechanisms,
+all components must propagate context down the chain.
+
+This imposes a difficulty because it cannot be assumed, and in many cases,
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:90]: This is a very long and nested sentence. It's a bit hard to follow for me. I'd prefer having more, but shorter sentences.

File: specification/trace/semantic_conventions/messaging.md [hunk 6]
```diff
@@ -80,6 +80,48 @@ Often such destinations are unnamed or have an auto-generated name.
...
+it is not even desired, that all components and layers to be instrumented and
+propagate context according to OpenTelemetry requirements.
+
+To be able to correlate consumer traces with producer traces without requiring
+intermediary instrumentation, the context needs to be propagated on a
+*per-message* basis instead of on a *per-request* basis. This allows all components
+to have access to the same per-message context information, making it possible
+to correlate all the stages involved in processing a message with the message's
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:98]:
```suggestion
to correlate all stages involved in processing a message with the message's
```

File: specification/trace/semantic_conventions/messaging.md [hunk 7]
```diff
@@ -80,6 +80,48 @@ Often such destinations are unnamed or have an auto-generated name.
...
+creation.
+
+A message *creation context* allows correlating producer with consumer(s)
+of a message and model the dependencies between them,
+regardless of the underlying messaging transport mechanism and its instrumentation.
+
+The message creation context is created by the producer and should be propagated
+to the consumer(s). Producer and consumer applications should be instrumented
+in a way so that the creation context is attached to messages and extracted
+from messages in a coordinated way.
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:108]: What does that mean? Do you mean a standardized way?

File: specification/trace/semantic_conventions/messaging.md [hunk 8]
```diff
@@ -80,6 +80,48 @@ Often such destinations are unnamed or have an auto-generated name.
...
+
+If the message creation context cannot be attached to the message and propagated,
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:110]:
```suggestion
If the message creation context cannot be attached and propagated with the message
```

Comment (pirgeo) [reply specification/trace/semantic_conventions/messaging.md:110]: I would flip the halves of the sentence around: 
`Consumer traces cannot be directly correlated to producer traces if the message creation context cannot be attached and propagated with the message.

File: specification/trace/semantic_conventions/messaging.md [hunk 9]
```diff
@@ -80,6 +80,48 @@ Often such destinations are unnamed or have an auto-generated name.
...
+consumer traces cannot be directly correlated to producer traces.
+
+A producer SHOULD attach a message creation context to each message.
+The message creation context SHOULD be attached in a way so that it is
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:114]:
```suggestion
The message creation context SHOULD be attached in such a way that it is
```

File: specification/trace/semantic_conventions/messaging.md [hunk 10]
```diff
@@ -80,6 +80,48 @@ Often such destinations are unnamed or have an auto-generated name.
...
+not possible to be changed by intermediaries.
+
+> This document does not specify the exact mechanisms on how the creation context
+is attached/extracted to/from messages. Future versions of these conventions
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:118]: I am not sure what the standard is on OTEL, do you prefix every line in a note with `>`? It doesn't change the rendered version, I think. Did you run `markdownlint`?

File: specification/trace/semantic_conventions/messaging.md [hunk 11]
```diff
@@ -80,6 +80,48 @@ Often such destinations are unnamed or have an auto-generated name.
...
+will give clear recommendations, following industry standards such as
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:119]:
```suggestion
will give clear recommendations, following industry standards including, but not limited to
```
Either is fine though, I think

File: specification/trace/semantic_conventions/messaging.md (hunk 10 above)
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md:118]: Good question. I always use `>` on the first line. Markdown lint didn't complain as well.

File: specification/trace/semantic_conventions/messaging.md (hunk 8 above)
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md:110]: Thanks, that's much better indeed!

File: specification/trace/semantic_conventions/messaging.md (hunk 5 above)
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md:90]: I agree. What you think about this:

```
This imposes a difficulty because it requires that all components are instrumented
and propagate context according to OpenTelemetry requirements. In reality,
this cannot be assumed, and in many cases is not even desired nor possible.
```

File: specification/trace/semantic_conventions/messaging.md (hunk 7 above)
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md:108]: I think what the original author of OTEP wanted to say is that producers and consumers must attach/extract the context at certain "points" during the message lifetime, in order to create a meaningful trace. For ex, producer would attach the context before sending, and a consumer would extract it upon receiving.. thus in a "coordinated" way.

File: specification/trace/semantic_conventions/messaging.md (hunk 5 above)
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md:90]: Or 

```
This imposes a requirement that all components are instrumented
and propagate context according to OpenTelemetry requirements. In reality,
this cannot be assumed, and in many cases is not even desired nor possible.
```

File: specification/trace/semantic_conventions/messaging.md (hunk 10 above)
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md]: I looked, it seems it's every line is prefixed in the examples I found. Changed in [4d89927](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/pull/22/commits/4d89927be29348fe8a06aee2193b170a3b6e7fc1)

File: specification/trace/semantic_conventions/messaging.md (hunk 5 above)
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md]: Went with the last in [4d89927](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/pull/22/commits/4d89927be29348fe8a06aee2193b170a3b6e7fc1). Let me know if it's better!

File: specification/trace/semantic_conventions/messaging.md (hunk 7 above)
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md:109]: I think I will remove this section: "Producer and consumer applications should be instrumented in a way [...]". It doesn't add any value and the exact way this will work is what the group is working on to define right now.

Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md]: Done [6c0c7c7](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/pull/22/commits/6c0c7c7b87a932d080561174c0a3141642373953)

File: specification/trace/semantic_conventions/messaging.md [hunk 12]
```diff
@@ -82,45 +83,42 @@ Given these definitions, the remainder of this section describes the semantic co
...
-A message may pass many different components and layers in one or more intermediaries
+A message may traverse many different components and layers in one or more intermediaries
 when it is propagated from the producer to the consumer(s). To be able to correlate
 consumer traces with producer traces using the existing context propagation mechanisms,
 all components must propagate context down the chain.
 
-This imposes a difficulty because it cannot be assumed, and in many cases,
-it is not even desired, that all components and layers to be instrumented and
-propagate context according to OpenTelemetry requirements.
+This imposes a requirement that all components are instrumented
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:91]: I am reluctant about the `This` here. You can't read this paragraph without the one before it. Maybe something like: The anatomy of systems relying on messaging imposes a requirement...?

Also imposes a *requirement* ... according to OpenTelemetry *requirements*.

File: specification/trace/semantic_conventions/messaging.md [hunk 13]
```diff
@@ -80,6 +81,45 @@ Often such destinations are unnamed or have an auto-generated name.
...
+### Context propagation
+
+A message may traverse many different components and layers in one or more intermediaries
+when it is propagated from the producer to the consumer(s). To be able to correlate
+consumer traces with producer traces using the existing context propagation mechanisms,
+all components must propagate context down the chain.
+
+This imposes a requirement that all components are instrumented
+and propagate context according to OpenTelemetry requirements. In reality,
+this cannot be assumed, and in many cases is not even desired or possible.
```
Comment (pirgeo) [inline specification/trace/semantic_conventions/messaging.md:93]:
```suggestion
this cannot be assumed, and in many cases is not desired or possible.
```

File: specification/trace/semantic_conventions/messaging.md (hunk 5 above)
Comment (pirgeo) [reply specification/trace/semantic_conventions/messaging.md]: Started a new conversation in https://github.com/dynatrace-oss-contrib/opentelemetry-specification/pull/22#discussion_r951387777. Still feels a bit clunky to me.

File: specification/trace/semantic_conventions/messaging.md (hunk 12 above)
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md]: Agreed. I changed it in [16bbdc7](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/pull/22/commits/16bbdc7d4447813da7e3209cd47437700fcd4a65).

File: specification/trace/semantic_conventions/messaging.md (hunk 13 above)
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/messaging.md]: Removed this sentence so not necessary anymore [16bbdc7](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/pull/22/commits/16bbdc7d4447813da7e3209cd47437700fcd4a65)

**Review by pirgeo — APPROVED** [review]
